zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandlerTest.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.exporter.handlers.operation;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.IncidentIntent;
+import io.camunda.zeebe.protocol.record.value.IncidentRecordValue;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class OperationFromIncidentHandlerTest extends AbstractOperationHandlerTest<IncidentRecordValue> {
```

Comment:
🔧 Personally I'm not a fan of using inheritance/abstractions in tests (also in general).

For tests it should be readable and understandable what they test and what they cover. If we have such a test class this is often not the case. (as I have to navigate to somewhere else to find the actual test).

Furthermore, we coupling these tests together. If we change now things in the abstract test, this might not apply for all other underneat and can cause trouble.

BTW the same applies for your actual inheritance, where I think we could also use composition instead. 

You could say the AbstractOperationHandler is actual a composable handler (call it `OperationHandler`).

This can be used by all the other classes, and they just delegate (use DI here as well). 

Within the tests you only need to test once `OperationHandler`, for the others you just need to test the differences like what they handle (check different intents). And in other cases they delegate to `OperationHandler` (could be tested with a mock)

Reply:
I will leave this up to you. I think it is not blocking, and we can adjust also if we ever need.

Reply:
I see what you mean, and I agree with the `composition over inheritance` argument in general 👍 
For this use-case, it's a very simple possible not ever touched (or extended) again piece of code, so I think we can live with it this way.

But for the sake of entertaining the point too, I'll try to fiddle around as see how it would look like if I change this to use composition. I'll push it if it seemed simple enough to finish in an hour. If it brings more complications, I'll just merge as is now.
WDYT?

Reply:
Sounds great thanks!

Reply:
I actually made the changes from inheritance to composition and changed the tests and everything. But I don't like it honestly... it's so much more boilerplate code for something that is very simple.

Reply:
<details>
<summary> check the patch below</summary>

```patch
Subject: [PATCH] test: add tests for OperationFromVariableDocumentHandler
---
Index: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandlerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandlerTest.java b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandlerTest.java
--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandlerTest.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandlerTest.java	(date 1730466200018)
@@ -8,21 +8,42 @@
 package io.camunda.exporter.handlers.operation;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
 
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.store.BatchRequest;
+import io.camunda.webapps.schema.entities.operation.OperationEntity;
 import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
 import io.camunda.zeebe.protocol.record.intent.IncidentIntent;
+import io.camunda.zeebe.protocol.record.intent.Intent;
 import io.camunda.zeebe.protocol.record.value.IncidentRecordValue;
+import io.camunda.zeebe.test.broker.protocol.ProtocolFactory;
 import java.util.stream.Stream;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-class OperationFromIncidentHandlerTest extends AbstractOperationHandlerTest<IncidentRecordValue> {
+class OperationFromIncidentHandlerTest {
+  private final ProtocolFactory factory = new ProtocolFactory();
+
+  private OperationFromIncidentHandler underTest;
+  private CompleteOperationHandler<IncidentRecordValue> completeOperationHandler;
 
   @BeforeEach
   void setUp() {
-    underTest = new OperationFromIncidentHandler(indexName);
-    valueType = ValueType.INCIDENT;
+    completeOperationHandler = mock(CompleteOperationHandler.class);
+    underTest = new OperationFromIncidentHandler(completeOperationHandler);
+  }
+
+  @Test
+  void testGetHandledValueType() {
+    assertThat(underTest.getHandledValueType()).isEqualTo(ValueType.INCIDENT);
+  }
+
+  @Test
+  void testGetEntityType() {
+    assertThat(underTest.getEntityType()).isEqualTo(OperationEntity.class);
   }
 
   @Test
@@ -38,4 +59,57 @@
         .map(this::generateRecord)
         .forEach(record -> assertThat(underTest.handlesRecord(record)).isFalse());
   }
+
+  @Test
+  void testCreateNewEntity() {
+    // given
+    final String id = "id";
+
+    // when
+    final var entity = underTest.createNewEntity(id);
+
+    // then
+    verify(completeOperationHandler).createNewEntity(id);
+  }
+
+  @Test
+  void testGenerateIds() {
+    // given
+    final var record = mock(Record.class);
+    // when
+    final var ids = underTest.generateIds(record);
+
+    // then
+    verify(completeOperationHandler).generateIds(record);
+  }
+
+  @Test
+  void testUpdateEntity() {
+    // given
+    final var record = mock(Record.class);
+    final var entity = new OperationEntity();
+
+    // when
+    underTest.updateEntity(record, entity);
+
+    // then
+    verify(completeOperationHandler).updateEntity(entity);
+  }
+
+  @Test
+  void shouldFlush() throws PersistenceException {
+    // given
+    final var entity = new OperationEntity();
+    final var batchRequest = mock(BatchRequest.class);
+
+    // when
+    underTest.flush(entity, batchRequest);
+
+    // then
+    verify(completeOperationHandler).flush(entity, batchRequest);
+  }
+
+  private Record<IncidentRecordValue> generateRecord(final Intent intent) {
+    return factory.generateRecord(ValueType.INCIDENT, r -> r.withIntent(intent));
+  }
 }
Index: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandler.java b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandler.java
--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandler.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromIncidentHandler.java	(date 1730466148144)
@@ -7,15 +7,28 @@
  */
 package io.camunda.exporter.handlers.operation;
 
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.handlers.ExportHandler;
+import io.camunda.exporter.store.BatchRequest;
+import io.camunda.webapps.schema.entities.operation.OperationEntity;
 import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
 import io.camunda.zeebe.protocol.record.intent.IncidentIntent;
 import io.camunda.zeebe.protocol.record.value.IncidentRecordValue;
+import java.util.List;
 
-public class OperationFromIncidentHandler extends AbstractOperationHandler<IncidentRecordValue> {
+public class OperationFromIncidentHandler
+    implements ExportHandler<OperationEntity, IncidentRecordValue> {
+
+  private final CompleteOperationHandler<IncidentRecordValue> completeOperationHandler;
 
   public OperationFromIncidentHandler(final String indexName) {
-    super(indexName);
+    completeOperationHandler = new CompleteOperationHandler<>(indexName);
+  }
+
+  protected OperationFromIncidentHandler(
+      final CompleteOperationHandler<IncidentRecordValue> completeOperationHandler) {
+    this.completeOperationHandler = completeOperationHandler;
   }
 
   @Override
@@ -23,8 +36,34 @@
     return ValueType.INCIDENT;
   }
 
+  @Override
+  public Class<OperationEntity> getEntityType() {
+    return OperationEntity.class;
+  }
+
   @Override
   public boolean handlesRecord(final Record<IncidentRecordValue> record) {
     return IncidentIntent.RESOLVED.equals(record.getIntent());
   }
+
+  @Override
+  public List<String> generateIds(final Record<IncidentRecordValue> record) {
+    return completeOperationHandler.generateIds(record);
+  }
+
+  @Override
+  public OperationEntity createNewEntity(final String id) {
+    return completeOperationHandler.createNewEntity(id);
+  }
+
+  @Override
+  public void updateEntity(final Record<IncidentRecordValue> record, final OperationEntity entity) {
+    completeOperationHandler.updateEntity(entity);
+  }
+
+  @Override
+  public void flush(final OperationEntity entity, final BatchRequest batchRequest)
+      throws PersistenceException {
+    completeOperationHandler.flush(entity, batchRequest);
+  }
 }
Index: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandler.java b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandler.java
--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandler.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandler.java	(date 1730465752676)
@@ -7,16 +7,27 @@
  */
 package io.camunda.exporter.handlers.operation;
 
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.handlers.ExportHandler;
+import io.camunda.exporter.store.BatchRequest;
+import io.camunda.webapps.schema.entities.operation.OperationEntity;
 import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
 import io.camunda.zeebe.protocol.record.intent.VariableDocumentIntent;
 import io.camunda.zeebe.protocol.record.value.VariableDocumentRecordValue;
+import java.util.List;
 
 public class OperationFromVariableDocumentHandler
-    extends AbstractOperationHandler<VariableDocumentRecordValue> {
+    implements ExportHandler<OperationEntity, VariableDocumentRecordValue> {
+  private final CompleteOperationHandler<VariableDocumentRecordValue> completeOperationHandler;
 
   public OperationFromVariableDocumentHandler(final String indexName) {
-    super(indexName);
+    completeOperationHandler = new CompleteOperationHandler<>(indexName);
+  }
+
+  protected OperationFromVariableDocumentHandler(
+      final CompleteOperationHandler<VariableDocumentRecordValue> completeOperationHandler) {
+    this.completeOperationHandler = completeOperationHandler;
   }
 
   @Override
@@ -24,8 +35,35 @@
     return ValueType.VARIABLE_DOCUMENT;
   }
 
+  @Override
+  public Class<OperationEntity> getEntityType() {
+    return OperationEntity.class;
+  }
+
   @Override
   public boolean handlesRecord(final Record<VariableDocumentRecordValue> record) {
     return VariableDocumentIntent.UPDATED.equals(record.getIntent());
   }
+
+  @Override
+  public List<String> generateIds(final Record<VariableDocumentRecordValue> record) {
+    return completeOperationHandler.generateIds(record);
+  }
+
+  @Override
+  public OperationEntity createNewEntity(final String id) {
+    return completeOperationHandler.createNewEntity(id);
+  }
+
+  @Override
+  public void updateEntity(
+      final Record<VariableDocumentRecordValue> record, final OperationEntity entity) {
+    completeOperationHandler.updateEntity(entity);
+  }
+
+  @Override
+  public void flush(final OperationEntity entity, final BatchRequest batchRequest)
+      throws PersistenceException {
+    completeOperationHandler.flush(entity, batchRequest);
+  }
 }
Index: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/AbstractOperationHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/AbstractOperationHandler.java b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/CompleteOperationHandler.java
rename from zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/AbstractOperationHandler.java
rename to zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/CompleteOperationHandler.java
--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/AbstractOperationHandler.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/CompleteOperationHandler.java	(date 1730461791737)
@@ -8,7 +8,6 @@
 package io.camunda.exporter.handlers.operation;
 
 import io.camunda.exporter.exceptions.PersistenceException;
-import io.camunda.exporter.handlers.ExportHandler;
 import io.camunda.exporter.store.BatchRequest;
 import io.camunda.webapps.schema.descriptors.operate.template.OperationTemplate;
 import io.camunda.webapps.schema.entities.operation.OperationEntity;
@@ -20,21 +19,14 @@
 import java.util.List;
 import java.util.Map;
 
-public abstract class AbstractOperationHandler<R extends RecordValue>
-    implements ExportHandler<OperationEntity, R> {
+public class CompleteOperationHandler<R extends RecordValue> {
 
   protected final String indexName;
 
-  public AbstractOperationHandler(final String indexName) {
+  public CompleteOperationHandler(final String indexName) {
     this.indexName = indexName;
   }
 
-  @Override
-  public Class<OperationEntity> getEntityType() {
-    return OperationEntity.class;
-  }
-
-  @Override
   public List<String> generateIds(final Record<R> record) {
     final long operationReference = record.getOperationReference();
     if (operationReference > 0) {
@@ -43,13 +35,11 @@
     return List.of();
   }
 
-  @Override
   public OperationEntity createNewEntity(final String id) {
     return new OperationEntity().setId(id);
   }
 
-  @Override
-  public void updateEntity(final Record<R> record, final OperationEntity entity) {
+  public void updateEntity(final OperationEntity entity) {
     entity
         .setState(OperationState.COMPLETED)
         .setLockOwner(null)
@@ -57,7 +47,6 @@
         .setCompletedDate(OffsetDateTime.now());
   }
 
-  @Override
   public void flush(final OperationEntity entity, final BatchRequest batchRequest)
       throws PersistenceException {
     final Map<String, Object> updateFields = new HashMap<>();
Index: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/AbstractOperationHandlerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/AbstractOperationHandlerTest.java b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/CompleteOperationHandlerTest.java
rename from zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/AbstractOperationHandlerTest.java
rename to zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/CompleteOperationHandlerTest.java
--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/AbstractOperationHandlerTest.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/CompleteOperationHandlerTest.java	(date 1730461922158)
@@ -8,7 +8,6 @@
 package io.camunda.exporter.handlers.operation;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 
@@ -18,30 +17,29 @@
 import io.camunda.webapps.schema.entities.operation.OperationEntity;
 import io.camunda.webapps.schema.entities.operation.OperationState;
 import io.camunda.zeebe.protocol.record.Record;
-import io.camunda.zeebe.protocol.record.RecordValue;
 import io.camunda.zeebe.protocol.record.ValueType;
-import io.camunda.zeebe.protocol.record.intent.Intent;
 import io.camunda.zeebe.test.broker.protocol.ProtocolFactory;
 import java.time.OffsetDateTime;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Stream;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
-abstract class AbstractOperationHandlerTest<R extends RecordValue> {
-  protected AbstractOperationHandler<R> underTest;
-  protected final ProtocolFactory factory = new ProtocolFactory();
-  protected final String indexName = OperationTemplate.INDEX_NAME;
-  protected ValueType valueType;
+class CompleteOperationHandlerTest {
+  private final ProtocolFactory factory = new ProtocolFactory();
 
-  @Test
-  void testGetHandledValueType() {
-    assertEquals(underTest.getHandledValueType(), valueType);
-  }
+  private final CompleteOperationHandler<?> underTest =
+      new CompleteOperationHandler<>(OperationTemplate.INDEX_NAME);
 
-  @Test
-  void testGetEntityType() {
-    assertThat(underTest.getEntityType()).isEqualTo(OperationEntity.class);
+  private static Stream<Arguments> valueTypes() {
+    return Stream.of(
+        Arguments.of(ValueType.INCIDENT),
+        Arguments.of(ValueType.PROCESS_INSTANCE),
+        Arguments.of(ValueType.VARIABLE_DOCUMENT));
   }
 
   @Test
@@ -57,11 +55,12 @@
     assertThat(entity.getId()).isEqualTo(id);
   }
 
-  @Test
-  void testGenerateIds() {
+  @ParameterizedTest()
+  @MethodSource("valueTypes")
+  void testGenerateIds(final ValueType valueType) {
     // given
     final long operationReference = 123L;
-    final Record<R> record =
+    final Record record =
         factory.generateRecord(valueType, r -> r.withOperationReference(operationReference));
     // when
     final List<String> ids = underTest.generateIds(record);
@@ -70,11 +69,12 @@
     assertThat(ids).containsExactly(String.valueOf(operationReference));
   }
 
-  @Test
-  void shouldNotGenerateId() {
+  @ParameterizedTest()
+  @MethodSource("valueTypes")
+  void shouldNotGenerateId(final ValueType valueType) {
     // given
     final long operationReference = -1;
-    final Record<R> record =
+    final Record record =
         factory.generateRecord(valueType, r -> r.withOperationReference(operationReference));
     // when
     final List<String> ids = underTest.generateIds(record);
@@ -85,11 +85,10 @@
   @Test
   void testUpdateEntity() {
     // given
-    final Record<R> record = factory.generateRecord(valueType);
     final OperationEntity entity = new OperationEntity();
 
     // when
-    underTest.updateEntity(record, entity);
+    underTest.updateEntity(entity);
 
     // then
     assertThat(entity.getState()).isEqualTo(OperationState.COMPLETED);
@@ -121,14 +120,6 @@
     underTest.flush(entity, mockRequest);
 
     // then
-    verify(mockRequest).update(indexName, entity.getId(), expectedUpdateFields);
-  }
-
-  protected Record<R> generateRecord(final Intent intent) {
-    return factory.generateRecord(valueType, r -> r.withIntent(intent));
-  }
-
-  protected Record<R> generateRecord(final Intent intent, final R value) {
-    return factory.generateRecord(valueType, r -> r.withIntent(intent).withValue(value));
+    verify(mockRequest).update(OperationTemplate.INDEX_NAME, entity.getId(), expectedUpdateFields);
   }
 }
Index: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandlerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandlerTest.java b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandlerTest.java
--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandlerTest.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromVariableDocumentHandlerTest.java	(date 1730465892416)
@@ -8,21 +8,42 @@
 package io.camunda.exporter.handlers.operation;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
 
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.store.BatchRequest;
+import io.camunda.webapps.schema.entities.operation.OperationEntity;
+import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
 import io.camunda.zeebe.protocol.record.intent.VariableDocumentIntent;
 import io.camunda.zeebe.protocol.record.value.VariableDocumentRecordValue;
+import io.camunda.zeebe.test.broker.protocol.ProtocolFactory;
 import java.util.stream.Stream;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-class OperationFromVariableDocumentHandlerTest
-    extends AbstractOperationHandlerTest<VariableDocumentRecordValue> {
+class OperationFromVariableDocumentHandlerTest {
+  private final ProtocolFactory factory = new ProtocolFactory();
+
+  private OperationFromVariableDocumentHandler underTest;
+  private CompleteOperationHandler<VariableDocumentRecordValue> completeOperationHandler;
 
   @BeforeEach
   void setUp() {
-    underTest = new OperationFromVariableDocumentHandler(indexName);
-    valueType = ValueType.VARIABLE_DOCUMENT;
+    completeOperationHandler = mock(CompleteOperationHandler.class);
+    underTest = new OperationFromVariableDocumentHandler(completeOperationHandler);
+  }
+
+  @Test
+  void testGetHandledValueType() {
+    assertThat(underTest.getHandledValueType()).isEqualTo(ValueType.VARIABLE_DOCUMENT);
+  }
+
+  @Test
+  void testGetEntityType() {
+    assertThat(underTest.getEntityType()).isEqualTo(OperationEntity.class);
   }
 
   @Test
@@ -38,4 +59,57 @@
         .map(this::generateRecord)
         .forEach(record -> assertThat(underTest.handlesRecord(record)).isFalse());
   }
+
+  @Test
+  void testCreateNewEntity() {
+    // given
+    final String id = "id";
+
+    // when
+    final var entity = underTest.createNewEntity(id);
+
+    // then
+    verify(completeOperationHandler).createNewEntity(id);
+  }
+
+  @Test
+  void testGenerateIds() {
+    // given
+    final var record = mock(Record.class);
+    // when
+    final var ids = underTest.generateIds(record);
+
+    // then
+    verify(completeOperationHandler).generateIds(record);
+  }
+
+  @Test
+  void testUpdateEntity() {
+    // given
+    final var record = mock(Record.class);
+    final var entity = new OperationEntity();
+
+    // when
+    underTest.updateEntity(record, entity);
+
+    // then
+    verify(completeOperationHandler).updateEntity(entity);
+  }
+
+  @Test
+  void shouldFlush() throws PersistenceException {
+    // given
+    final var entity = new OperationEntity();
+    final var batchRequest = mock(BatchRequest.class);
+
+    // when
+    underTest.flush(entity, batchRequest);
+
+    // then
+    verify(completeOperationHandler).flush(entity, batchRequest);
+  }
+
+  private Record<VariableDocumentRecordValue> generateRecord(final Intent intent) {
+    return factory.generateRecord(ValueType.VARIABLE_DOCUMENT, r -> r.withIntent(intent));
+  }
 }
Index: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandlerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandlerTest.java b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandlerTest.java
--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandlerTest.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandlerTest.java	(date 1730466100515)
@@ -9,23 +9,44 @@
 
 import static io.camunda.exporter.handlers.operation.OperationFromProcessInstanceHandler.ELIGIBLE_STATES;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
 
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.store.BatchRequest;
+import io.camunda.webapps.schema.entities.operation.OperationEntity;
+import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
 import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
 import io.camunda.zeebe.protocol.record.value.BpmnElementType;
 import io.camunda.zeebe.protocol.record.value.ImmutableProcessInstanceRecordValue;
 import io.camunda.zeebe.protocol.record.value.ProcessInstanceRecordValue;
+import io.camunda.zeebe.test.broker.protocol.ProtocolFactory;
 import java.util.stream.Stream;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-class OperationFromProcessInstanceHandlerTest
-    extends AbstractOperationHandlerTest<ProcessInstanceRecordValue> {
+class OperationFromProcessInstanceHandlerTest {
+  private final ProtocolFactory factory = new ProtocolFactory();
+
+  private OperationFromProcessInstanceHandler underTest;
+  private CompleteOperationHandler<ProcessInstanceRecordValue> completeOperationHandler;
 
   @BeforeEach
   void setUp() {
-    underTest = new OperationFromProcessInstanceHandler(indexName);
-    valueType = ValueType.PROCESS_INSTANCE;
+    completeOperationHandler = mock(CompleteOperationHandler.class);
+    underTest = new OperationFromProcessInstanceHandler(completeOperationHandler);
+  }
+
+  @Test
+  void testGetHandledValueType() {
+    assertThat(underTest.getHandledValueType()).isEqualTo(ValueType.PROCESS_INSTANCE);
+  }
+
+  @Test
+  void testGetEntityType() {
+    assertThat(underTest.getEntityType()).isEqualTo(OperationEntity.class);
   }
 
   @Test
@@ -63,4 +84,59 @@
         .map(i -> generateRecord(i, value))
         .forEach(record -> assertThat(underTest.handlesRecord(record)).isFalse());
   }
+
+  @Test
+  void testCreateNewEntity() {
+    // given
+    final String id = "id";
+
+    // when
+    final var entity = underTest.createNewEntity(id);
+
+    // then
+    verify(completeOperationHandler).createNewEntity(id);
+  }
+
+  @Test
+  void testGenerateIds() {
+    // given
+    final var record = mock(Record.class);
+    // when
+    final var ids = underTest.generateIds(record);
+
+    // then
+    verify(completeOperationHandler).generateIds(record);
+  }
+
+  @Test
+  void testUpdateEntity() {
+    // given
+    final var record = mock(Record.class);
+    final var entity = new OperationEntity();
+
+    // when
+    underTest.updateEntity(record, entity);
+
+    // then
+    verify(completeOperationHandler).updateEntity(entity);
+  }
+
+  @Test
+  void shouldFlush() throws PersistenceException {
+    // given
+    final var entity = new OperationEntity();
+    final var batchRequest = mock(BatchRequest.class);
+
+    // when
+    underTest.flush(entity, batchRequest);
+
+    // then
+    verify(completeOperationHandler).flush(entity, batchRequest);
+  }
+
+  protected Record<ProcessInstanceRecordValue> generateRecord(
+      final Intent intent, final ProcessInstanceRecordValue value) {
+    return factory.generateRecord(
+        ValueType.PROCESS_INSTANCE, r -> r.withIntent(intent).withValue(value));
+  }
 }
Index: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandler.java b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandler.java
--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandler.java	(revision 22e31a55df21b6ad5835bdb0c125ae0910705aed)
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/operation/OperationFromProcessInstanceHandler.java	(date 1730466153800)
@@ -10,19 +10,30 @@
 import static io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent.ELEMENT_MIGRATED;
 import static io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent.ELEMENT_TERMINATED;
 
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.handlers.ExportHandler;
+import io.camunda.exporter.store.BatchRequest;
+import io.camunda.webapps.schema.entities.operation.OperationEntity;
 import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.ValueType;
 import io.camunda.zeebe.protocol.record.intent.Intent;
 import io.camunda.zeebe.protocol.record.value.BpmnElementType;
 import io.camunda.zeebe.protocol.record.value.ProcessInstanceRecordValue;
+import java.util.List;
 import java.util.Set;
 
 public class OperationFromProcessInstanceHandler
-    extends AbstractOperationHandler<ProcessInstanceRecordValue> {
+    implements ExportHandler<OperationEntity, ProcessInstanceRecordValue> {
   protected static final Set<Intent> ELIGIBLE_STATES = Set.of(ELEMENT_MIGRATED, ELEMENT_TERMINATED);
+  private final CompleteOperationHandler<ProcessInstanceRecordValue> completeOperationHandler;
 
   public OperationFromProcessInstanceHandler(final String indexName) {
-    super(indexName);
+    completeOperationHandler = new CompleteOperationHandler<>(indexName);
+  }
+
+  protected OperationFromProcessInstanceHandler(
+      final CompleteOperationHandler<ProcessInstanceRecordValue> completeOperationHandler) {
+    this.completeOperationHandler = completeOperationHandler;
   }
 
   @Override
@@ -30,11 +41,38 @@
     return ValueType.PROCESS_INSTANCE;
   }
 
+  @Override
+  public Class<OperationEntity> getEntityType() {
+    return OperationEntity.class;
+  }
+
   @Override
   public boolean handlesRecord(final Record<ProcessInstanceRecordValue> record) {
     return isProcessEvent(record.getValue()) && ELIGIBLE_STATES.contains(record.getIntent());
   }
 
+  @Override
+  public List<String> generateIds(final Record<ProcessInstanceRecordValue> record) {
+    return completeOperationHandler.generateIds(record);
+  }
+
+  @Override
+  public OperationEntity createNewEntity(final String id) {
+    return completeOperationHandler.createNewEntity(id);
+  }
+
+  @Override
+  public void updateEntity(
+      final Record<ProcessInstanceRecordValue> record, final OperationEntity entity) {
+    completeOperationHandler.updateEntity(entity);
+  }
+
+  @Override
+  public void flush(final OperationEntity entity, final BatchRequest batchRequest)
+      throws PersistenceException {
+    completeOperationHandler.flush(entity, batchRequest);
+  }
+
   protected boolean isProcessEvent(final ProcessInstanceRecordValue recordValue) {
     return isOfType(recordValue, BpmnElementType.PROCESS);
   }

```
</details>

Reply:
ok :) Thanks for trying it out 👍🏼